### PR TITLE
🚀🐟 + 🌊 | Spacefish swim towards v1.0

### PR DESCRIPTION
**Pull request by Snuggle**
## Description

This pull request attempts to help push Spacefish towards complete feature parity with Spaceship with a 🌊

## Motivation and Context

It aims to fix/implement the following issues: #20 #28 #29 #32 #34 #37 and help push Spacefish to [the v1.0 milestone](https://github.com/matchai/spacefish/milestone/1).

## Types of changes

- [x]  Several new features (non-breaking changes which adds functionality)

## How Has This Been Tested?

- [x] I have tested using **MacOS**
- [x] I have tested using **Linux**

## Default Pull Request Checklist:

- [x] I have checked that no other PR duplicates mine
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

## Personal Development Checklist:
- [x] Create initial rough translation from Zsh to Fish.
- [x] Write a lot of documentation.
- [ ] Manually test Windows battery-acpi on a Windows laptop.
- [x] Manually test pyenv section.
- [x] Manually test background jobs section.
- [x] Manually test time section.
- [x] Manually test exit code section.
- [x] Manually tested everything.
- [x] Add automated tests.
- [x] Automated tests complete without errors.
- [x] Swim-to-v1.0 is ready to merge. 🚀🐟 + 🌊

### CODE REVIEW

**Comment by matchai** [comment]
Woah! Looks like some serious work is being done here! 👏👏👏

I would suggest breaking every new section into its own PR with its milestone set to `1.0.0` to make the reviewing merging process cleaner.

Looking forward to reaching 1.0.0! 🎉

**Comment by Snuggle** [comment]
To clarify, split each section relative to this branch, or from the master branch? A lot of the sections have been written in commit 41847d0ef2ccb77ddd23d2e42f77403efb9af6db, but need testing, documentation and possibly small fixes.

I'm also not able to set milestones on issues or pull requests, I think only maintainers/owners can do that. :yum:

File: functions/__sf_section_exit_code.fish
Comment (matchai) [inline functions/__sf_section_exit_code.fish]: I would suggest using [`$sf_exit_code`](https://github.com/matchai/spacefish/blob/master/fish_prompt.fish#L3) (which was used [in char](https://github.com/matchai/spacefish/blob/master/functions/__sf_section_char.fish#L24))
`$status` will have the status code of the last command, which in thais case would be the last `__sf_util_set_default`. 😉

**Comment by matchai** [comment]
I was thinking that we would have each section be its own PR relative to master, but if we're already far gone, it'd be something to keep in mind for next time. 😄

**Comment by Snuggle** [comment]
I've noticed quite a bit of the documentation is out of order, I'll shuffle things around to make it in the same order as the prompt, for ease-of-use. :slightly_smiling_face:

File: functions/__sf_section_jobs.fish
Comment (Snuggle) [inline functions/__sf_section_jobs.fish]: This line in particular seems to raise issues on MacOS.

**Comment by Snuggle** [comment]
Heya! This is almost ready to merge, this just needs testing for MacOS, which I'm not able to do. The `AppVeyor` and `Travis-CI, Linux` build passes, but there is an issue on MacOS.

image

Could you please test the line above on MacOS and modify it, because it appears to have a bunch of random spaces which isn't the case on Linux. It's supposed to just return the number of background jobs/processes.

Other than that, I'd love to hear if you can spot anything I've missed. :slightly_smiling_face:

File: docs/Options.md
Comment (matchai) [inline docs/Options.md]: A few of these headers should be formatted as they are in spaceship:
> ### Execution time (`exec_time`)

File: functions/__sf_section_exit_code.fish
Comment (matchai) [inline functions/__sf_section_exit_code.fish]: Would suggest removing double-quotes around the value if there's no whitespace 😄

File: functions/__sf_section_exit_code.fish
Comment (matchai) [inline functions/__sf_section_exit_code.fish:1]: It looks like this file and a few of the other section files are using spaces rather than tabs. Just a heads up! 🤠

File: functions/__sf_section_jobs.fish
Comment (matchai) [inline functions/__sf_section_jobs.fish]: A couple more variables with quotes.
Stick to quotes around empty strings and variables with whitespace.

File: tests/__sf_section_jobs.test.fish
Comment (matchai) [inline tests/__sf_section_jobs.test.fish]: Instead of creating a function, you can mock it using the mock library after sourcing it (`source $DIRNAME/mock.fish`):
```
mock killall 0 "sleep 5"
```

File: functions/__sf_section_jobs.fish
Comment (matchai) [inline functions/__sf_section_jobs.fish]: The tests pass just fine on the fish installation on my Mac.
I would try wrapping this variable in quotes since I've seen some strange behavior when concatenating variables without quotes. That might be the cause of the failing Travis builds.

**Comment by Snuggle** [comment]
The MacOS issue still occurs, it doesn't seem to be an issue with quotation marks.

It looks like either `jobs | wc -l` is outputting whitespace, which it *shouldn't* or there is some issue with how MacOS concatenates the `✦` symbol.

**Comment by matchai** [comment]
@Snuggle any chance that we could remove jobs from this PR and take care of it separately? The rest of it looks like it's ready to be reviewed.

File: docs/Options.md
Comment (matchai) [inline docs/Options.md]: I know this line was carried over from spaceship, but I'd suggest fixing up the English a little. 😅

File: docs/README.md
Comment (matchai) [inline docs/README.md]: Looks like you're missing a closing parenthesis here.

File: functions/__sf_section_battery.fish
Comment (matchai) [inline functions/__sf_section_battery.fish]: Looks good to me. I would just suggest removing the issue reference here.

File: functions/__sf_section_exit_code.fish
Comment (matchai) [inline functions/__sf_section_exit_code.fish]: I would suggest wrapping concatenations like this in quotes.
It appears that in fish, if one of the two values is null, the concatenated value is null. If the exit code symbol, in this case, is set to nothing, the exit code won't appear. 👀

File: functions/__sf_section_jobs.fish
Comment (matchai) [inline functions/__sf_section_jobs.fish]: To keep consistent with other strings, we should remove the quotes around `blue` 😄

File: functions/__sf_section_pyenv.fish
Comment (matchai) [inline functions/__sf_section_pyenv.fish]: Same string note as above ☝️

File: functions/__sf_section_pyenv.fish
Comment (matchai) [inline functions/__sf_section_pyenv.fish]: Same concatenation note as above ☝️

File: functions/__sf_section_jobs.fish
Comment (matchai) [inline functions/__sf_section_jobs.fish]: It looks like most uses of `__sf_lib_section` aren't indented after the line of the function call. I do prefer how it looks indented.

Would you mind indenting the other uses of `__sf_lib_section` in the other functions as part of your PR?

File: tests/__sf_section_time.test.fish
Comment (matchai) [inline tests/__sf_section_time.test.fish]: Looks like the contents of the fishtest parentheses are indented in the other uses.
I would suggest indenting these as well.

File: functions/__sf_section_pyenv.fish
Comment (matchai) [inline functions/__sf_section_pyenv.fish]: Looks like this file is still using spaces for indentation.

File: tests/__sf_section_time.test.fish
Comment (matchai) [inline tests/__sf_section_time.test.fish]: Seems like there's a couple spaces mixed in with the tabs here.

**Comment by Snuggle** [comment]
How do you pass args into a mock function? I'm trying to fix my tests since they are based on the current time and the tests can fail if the test takes >1 second to complete (See most recent AppVeyor failure), but I need to pass both the formatting string and set the current time manually.
image

**Comment by matchai** [comment]
I'm not sure I totally understand your question.

What is shown in your screenshot should mock the `newdate` function with an exit code of 0, and should run `date --date=@2147483647 $args`, replacing `$args` with its assigned value.

**Comment by Snuggle** [comment]
image
I'm trying to fix this bug and I need to call a mock date function that has a set, static time. NOT the current time. But I would need both the date format and the static time as an argument.

`mock newdate 0 "date --date=@2147483647"`

`set time_str "$time_str"(newdate +%I:%M:%S) `

I would like the mock function to return the value of `date --date=@2147483647 +%I:%M:%S`, but I can't figure out how to pass it. At the moment it returns the value of `date --date=@2147483647`.

**Comment by matchai** [comment]
image

Works fine for me 😄

**Comment by Snuggle** [comment]
Ah, but I need to do `newdate "+%I:%M:%S"`. The time format changes and cannot be _in_ the mock function. It would have to be passed.

**Comment by matchai** [comment]
In that case you'll probably want to write a function.

image

**Comment by Snuggle** [comment]
Gosh, trying to get things working in MacOS through automated testing is a real struggle. I need to fire up a MacOS virtual machine instead of spamming commits.

**Comment by Snuggle** [comment]
MacOS why?!  (╯ರ ~ ರ）╯彡 ┻━┻
Time to rewrite a part of this for MacOS's BSD date command.
image

File: functions/__sf_section_time.fish
Comment (matchai) [inline functions/__sf_section_time.fish]: You should be able to replace the two repeated cases with a single case with multiple arguments:
```fish
switch (uname) 
	case 'Darwin' '*BSD' 
		# MacOS + BSD use their own "date" coreutil.
		set time_format "-u" "-r 1536116421" "$time_format"
	case '*'
		set time_format "-u" "-d @1536116421" "$time_format"
end
```

Comment (matchai) [reply functions/__sf_section_time.fish]: I personally installed `coreutils` on mac to get the GNU utils instead of the existing BSD ones. Maybe instead of checking uname, we should use a defining property of the util. https://stackoverflow.com/a/8748344

The following worked for me:
```fish
date --version >/dev/null
  switch $status
      case 0
          command date "-u" "-d @1536116421"
      case '*'
          command date "-u" "-r 1536116421"
  end
```

Comment (matchai) [reply functions/__sf_section_time.fish]: I think I've found a slightly more elegant testing strategy for `__sf_section_time`:

Having the following function in the test file allows us to mock date with the appropriate GNU/BSD version:

```fish
function date -a time_format
	command date --version >/dev/null
	switch $status
		case 0
			command date "-u" "-d @1536116421" "$time_format"
		case '*'
			command date "-u" "-r 1536116421" "$time_format"
	end
end
```

With this, we can replace any `time_date` calls with `date`, removing any test code from the implementation 😄

Comment (matchai) [inline functions/__sf_section_time.fish]: I personally installed `coreutils` on my macos to replace the BSD utils with GNU utils. Instead of checking uname, maybe we should use something else to identify which is being used: https://stackoverflow.com/a/8748344

Comment (matchai) [reply functions/__sf_section_time.fish]: I think I found a more elegant solution for testing `__sf_section_time`:

By having this function in the test file, we will be able to use `date` instead of `time_date`, removing any test code from our implementation. 😄 

```fish
function date -a time_format
	command date --version >/dev/null
	switch $status
		case 0
			command date "-u" "-d @1536116421" "$time_format"
		case '*'
			command date "-u" "-r 1536116421" "$time_format"
	end
end
```

Comment (Snuggle) [reply functions/__sf_section_time.fish]: Oh, this is possible? I wouldn't have thought MacOS would let you change to GNU coreutils. This is a much better way to do it, since it relies on the command itself instead of the OS.

This might also explain why you couldn't reproduce the `jobs` issue from earlier, huh.

File: tests/__sf_section_time.test.fish
Comment (matchai) [inline tests/__sf_section_time.test.fish]: Looks like the Travis logs on macos are complaining about `--version`.
I guess we'll want to redirect both stdout and stderr.
```fish
command date --version >/dev/null ^/dev/null
```

File: tests/__sf_section_time.test.fish
Comment (matchai) [inline tests/__sf_section_time.test.fish:17]: I would suggest we add a teardown function:

```fish
function teardown
	functions --erase date
end
```

**Comment by matchai** [comment]
Looks ready to merge after those last couple comments 👍
Thank you for taking the time to work on this monster PR! 🎉

**Comment by Snuggle** [comment]
No problem! It was a pleasure finding a project that I loved _and_ could understand. Open-source can feel really daunting when you're faced with codebases that are monstrously big. Thanks for the thorough reviews and for porting my favourite Zsh prompt theme to Fish! I **_love_** the name and aesthetic, too.

(I also need to fix Atom because it's throwing borked whitespace all over the place.)

**Review by matchai — APPROVED** [review]

**Comment by matchai** [comment]
I'm glad you found this project to be easy to contribute to!
I really appreciate your kind words. 😄

**Pull request merged**
